examples/tcp-retransmit.bpf.c

```diff
@@ -0,0 +1,88 @@
+#include <vmlinux.h>
+#include <bpf/bpf_tracing.h>
+#include "maps.bpf.h"
+#include <bpf/bpf_core_read.h>
+
+#define MAX_ENTRIES 8192
+#define RETRANSMIT  1
+#define TLP         2
+
+#define AF_INET		2
+#define AF_INET6	10
+
+struct ipv4_key_t {
+	u32 saddr;
+	u32 daddr;
+	__u16 sport;
+	__be16 dport;
+    int type;
+};
+
+struct ipv6_key_t {
+	u8 saddr[16];
+	u8 daddr[16];
+	__be16 sport;
+	__u16 dport;
+    int type;
+};
+
+struct {
+	__uint(type, BPF_MAP_TYPE_HASH);
+	__uint(max_entries, MAX_ENTRIES);
+	__type(key, struct ipv4_key_t);
+	__type(value, u64);
+} tcp_retransmit_ipv4_count SEC(".maps");
+
+struct {
+	__uint(type, BPF_MAP_TYPE_HASH);
+	__uint(max_entries, MAX_ENTRIES);
+	__type(key, struct ipv6_key_t);
+	__type(value, u64);
+} tcp_retransmit_ipv6_count SEC(".maps");
+
+static int trace_event(void *ctx, const struct sock *skp, int type)
+{
+	__u32 family;
+	family = BPF_CORE_READ(skp, __sk_common.skc_family);
```

Comment:
No reason to split this into two lines, you can do:

```
u32 family = BPF_CORE_READ(skp, __sk_common.skc_family);
```

With `fentry` instead of `kprobe` you should also be able to drop `BPF_CORE_READ` and use normal member access.

Reply:
Did I understand correctly that you are suggesting using `kfunc` instead of `kprobe`?
Do you have a similar implementation at hand?

Reply:
Look for `fentry` in other examples to see how it's used:

* https://github.com/cloudflare/ebpf_exporter/blob/998f7f5260/examples/cfs-throttling.bpf.c#L22
* https://github.com/cloudflare/ebpf_exporter/blob/998f7f5260/examples/kstack.bpf.c#L21

I avoid using `kfunc` because it means multiple things:

* https://docs.kernel.org/bpf/kfuncs.html
* https://github.com/iovisor/bpftrace/blob/3afd12a9de/src/bpffeature.cpp#L634

Despite bpftrace calling this kfuncs, these are fprobes.